gm4_soul_fragments/data/gm4_soul_fragments/functions/main.mcfunction

```diff
@@ -0,0 +1,11 @@
+#runs stuff on death/respawn
+execute as @a[scores={gm4_sf_died=1..}] at @s run function gm4_soul_fragments:player/died
+execute as @e[type=player,scores={gm4_sf_rspwn=1..}] at @s run function gm4_soul_fragments:player/revived
```

Comment:
As far as I understand it you're abusing the way @a and @e function here to run one of these after respawn and one before respawn.

- First of all, you should probably get rid of one of the death scoreboards and give the player a dummy score once you detect them dying. You should then base your respawn test on that dummy score rather than a death score. This will ensure that your respawn commands only run on players that actually dropped a "soul fragment" and will get rid of one of the two death scoreboards.
- You could also move one of these away from main and use an advancement to trigger the respective function. I've been playing around with that lately and what you'd want is `"trigger": "minecraft:entity_hurt_player"` to call a function that then checks their death score for being 1 or more and resets said score. This would effectively make the death detection clockless (I think that would run right on death, at the location the player died at).